src/util/address.rs

```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
             "p2sh" => Ok(AddressType::P2sh),
             "p2wpkh" => Ok(AddressType::P2wpkh),
             "p2wsh" => Ok(AddressType::P2wsh),
+            "p2tr" => Ok(AddressType::P2tr),
             _ => Err(()),
         }
     }
 }
 
+/// Version of the WitnessProgram: first byte of `scriptPubkey` in
+/// transaction output for transactions starting with opcodes ranging from 0
+/// to 16 (inclusive).
+///
+/// Structure helps to limit possible version of the witness according to the
+/// specification; if a plain `u8` type will be used instead it will mean that
+/// version > 16, which is incorrect.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum WitnessVersion {
+    /// Current, initial version of Witness Program. Used for P2WPKH and P2WPK
+    /// outputs
+    V0,
+    /// Forthcoming second version of Witness Program, which (most probably)
+    /// will be used for Taproot
+    V1,
+    /// Future (unsupported) version of Witness Program
+    V2,
+    /// Future (unsupported) version of Witness Program
+    V3,
+    /// Future (unsupported) version of Witness Program
+    V4,
+    /// Future (unsupported) version of Witness Program
+    V5,
+    /// Future (unsupported) version of Witness Program
+    V6,
+    /// Future (unsupported) version of Witness Program
+    V7,
+    /// Future (unsupported) version of Witness Program
+    V8,
+    /// Future (unsupported) version of Witness Program
+    V9,
+    /// Future (unsupported) version of Witness Program
+    V10,
+    /// Future (unsupported) version of Witness Program
+    V11,
+    /// Future (unsupported) version of Witness Program
+    V12,
+    /// Future (unsupported) version of Witness Program
+    V13,
+    /// Future (unsupported) version of Witness Program
+    V14,
+    /// Future (unsupported) version of Witness Program
+    V15,
+    /// Future (unsupported) version of Witness Program
+    V16,
+}
+
+impl Display for WitnessVersion {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let ver = u8::from_str(s).map_err(|_| Error::UnparsableWitnessVersion)?;
```

Comment:
This loses information from the inner error. :(

I strongly believe errors should not discard valuable information like this. The variant should contain the inner error.

Reply:
This would require another error variant, but it might be the right thing to do depending on the use case (I'm not sure why `FromStr` and `Display` even exist for a witness version, it's something purely internal imo, but it didn't seem to hurt)